src/main/java/io/neonbee/internal/codec/DataExceptionMessageCodec.java

```diff
@@ -0,0 +1,117 @@
+            encoded = Buffer.buffer((String) obj);
+        } else if (obj instanceof JsonObject) {
+            buffer.appendInt(1); // 1 means JsonObject
+            encoded = ((JsonObject) obj).toBuffer();
+        } else if (obj instanceof JsonArray) {
+            buffer.appendInt(2); // 2 means JsonArray
+            encoded = ((JsonArray) obj).toBuffer();
+        } else {
+            throw new UnsupportedOperationException(
+                    "Serialization of type " + Object.class.getName() + " is not supported");
+        }
+        buffer.appendInt(encoded.length());
+        buffer.appendBuffer(encoded);
+    }
+
+    @Override
+    public DataException decodeFromWire(int pos, Buffer buffer) {
+        int temporaryPosition = pos;
+        String message = null;
+        String failureDetailKey = null;
+        Object failureDetailValue = null;
+        int length;
+        Map<String, Object> failureDetail = Map.of();
+        // decode failure code
+        int failureCode = buffer.getInt(temporaryPosition);
+        // decode message
+        temporaryPosition += Integer.BYTES * 2;
+        if (temporaryPosition <= buffer.length()) {
+            length = buffer.getInt(temporaryPosition);
+            temporaryPosition += Integer.BYTES;
+            if (length > 0) {
+                message = buffer.slice(temporaryPosition, temporaryPosition + length).toString();
+                temporaryPosition += length;
+            }
+        }
+        // decode failureDetail
+        temporaryPosition += Integer.BYTES;
+        if (temporaryPosition <= buffer.length()) {
+            length = buffer.getInt(temporaryPosition);
+            temporaryPosition += Integer.BYTES;
+            if (length > 0) {
+                failureDetailKey = buffer.slice(temporaryPosition, temporaryPosition + length).toString();
+                temporaryPosition += length;
+                int objType = buffer.getInt(temporaryPosition);
+                temporaryPosition += Integer.BYTES;
+                length = buffer.getInt(temporaryPosition);
+                temporaryPosition += Integer.BYTES;
+                Buffer failureDetailBodyAsBuffer = buffer.slice(temporaryPosition, temporaryPosition + length);
+                switch (objType) {
+                case 0:
+                    failureDetailValue = failureDetailBodyAsBuffer.toString();
+                    break;
+                case 1:
+                    failureDetailValue = new JsonObject(failureDetailBodyAsBuffer);
+                    break;
+                case 2:
+                    failureDetailValue = new JsonArray(failureDetailBodyAsBuffer);
+                    break;
+                default:
+                    throw new UnsupportedOperationException(
+                            "Deserialization of type " + Object.class.getName() + " is not supported");
+                }
+            }
+            failureDetail = Map.of(failureDetailKey, failureDetailValue);
+        }
```

Comment:
```java
int tmpPos = pos; // such a stupid rule that we have to put this into an own variable here, if we have a perfectly good inbound variable already ;)
int failureCode = buffer.getInt(tmpPos);
tmpPos += 4; // this is how Vert.x does it... think we should do it the same, see https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/eventbus/impl/codecs/ReplyExceptionMessageCodec.java#L44

int strLength = buffer.getInt(tmpPos);
tmpPos += 4;
String message = strLength == -1 ? null : new String(buffer.getBytes(tmpPos, tmpPos + strLength), CharsetUtil.UTF_8);
tmpPos += strLength;

int detailLength = buffer.getInt(tmpPos);
Map<String, Object> failureDetail = buffer.getBuffer(tmpPos, tmpPos + detailLength).toJsonObject().getMap();

return new DataException(failureCode, message, failureDetail);
```

Reply:
Hi Kristian,

thanks for the concrete proposal! I have implemented it and there seems to be one issue:
(according to my understanding) when decoding the failureDetail from the buffer, in case the value of the map is i.e. JsonObject, it get decoded to a LinkedHashMap instead of a JsonObject. The same for a JsonArray -> decoded to an ArrayList. Basically I noticed that the unit tests were failing when in the map I had a JsonObject or a JsonArray. 

The equals method of the DataException checks that the equality between 2 objects, with the default equals method of the Map -  I would say this is right, or?

I have attached a screenshot of the actual and expected objects from the failing test to clarify what I summed up above. 

<img width="1659" alt="Screenshot 2022-08-25 at 21 44 40" src="https://user-images.githubusercontent.com/2833188/186769602-36f84cac-9bea-46d5-9f28-76a37802cd09.png">

Reply:
Mhm yes I see the problem... There is no "real" solution for this because internally JsonObject is handled as Map and JsonArray is handled as List... and JsonObject/Array has handling in the getter method for that. So what it was "originally" can not be determined. As in my humble oppinion we can assume that most people will be dealing with JsonObject / JsonArray rather than with plain lists etc. for the failureDetail, my suggestion would be to use the stream method instead, when decoding the result, so instead of doing:

```java
buffer.getBuffer(tmpPos, tmpPos + detailLength).toJsonObject().getMap();
```

Do a:

```java
buffer.getBuffer(tmpPos, tmpPos + detailLength).toJsonObject().stream().collect(Collectors.toMap(Entry::getKey, Entry::getValue));
```

That should trigger the internal conversion by Vert.x into JsonObject / JsonArray. Hope this helps! :)